lib/engine/game/g_18_norway/steps/token.rb

```diff
@@ -21,24 +36,48 @@ def can_place_token?(_entity)
           def process_place_token(action)
             entity = action.entity
             city = action.city
-            connected_city = @game.loading || @game.token_graph_for_entity(entity).connected_nodes(entity)[city]
-            place_token(entity, action.city, action.token) if connected_city
+            hex = city.hex
+            connected_city = @game.token_graph_for_entity(entity).connected_nodes(entity)[city]
+            if connected_city
+              place_token(entity, action.city, action.token)
+            elsif @game.harbor_hex?(city.hex) && !connected_city
+              city_string = city.hex.tile.cities.size > 1 ? " city #{city.index}" : ''
+              city_name = @game.harbor_city_id_by_harbor_id(hex.id) + city_string
+              raise GameError, "Cannot reach city #{city_name}" unless check_available_harbour(entity, hex)
+
+              place_token(entity, action.city, action.token, connected: false)
 
-            unless connected_city
-              @game.abilities(entity, :token) do |ability|
-                place_token(entity, action.city, action.token, connected: false, special_ability: ability)
-              end
+            else
+              city_string = hex.tile.cities.size > 1 ? " city #{city.index}" : ''
+              raise GameError, "Cannot place token on #{hex.name}#{city_string} because it is not connected"
             end
             @game.clear_graph
             pass!
           end
 
-          def check_connected(entity, city, hex)
-            return if @game.loading || @game.token_graph_for_entity(entity).connected_nodes(entity)[city]
-            return if @game.loading || @game.ferry_graph.connected_nodes(entity)[city]
+          def place_token(entity, city, token, connected: true, extra_action: false,
```

Comment:
I don't like all this code cut'n'pasted into this method. I'm not sure I've understood the rules correctly, but is the code to allow placing port tokens in the sea hexes with the port symbols? It might be possible to have this as hex tokens, so instead of having to reimplement this entire method just to set `extra_slot`, you could do something like:
```
          def place_token(entity, city, token, connected: true, extra_action: false,
                          special_ability: nil, check_tokenable: true, spender: nil, same_hex_allowed: false)
            if @game.harbor_hex?(hex)
              # will need to check connection and pay cost
              hex.place_token(hex)
            else
              super
            end
          end
```

Reply:
I have tried to find another way to do this but have not found something that seems to be better so if you find another way then I am all for it.

One problem that might occur with hex_token is that it might be very crowded on the the hex.
Since all companies are allowed to put a harbor token in the harbor as long as they can trace to it.

The rules are that you are using the harbor tokens to "open" up for ships to trace a route through the port. The harbor token does not have any effect on the trains.

You can trace to a harbor either by train to the port or via the ships.

Reply:
![image](https://github.com/user-attachments/assets/af3cd4dd-5c14-499f-9285-3ca7a309ae1a)

Reply:
OK, that makes sense.
Instead of copying most of the `place_token` code into the 18Norway code, it might be better to add an `extra_slot` named parameter to `Engine::Step::Tokener.place_token` with a default value of `false`.

Reply:
@patrikolesen Have you had a chance to look at making this change?

Reply:
Sorry got caught in day work
But I finally got around to it and I agree that was a good improvment

Should I rebase to latest master?

Reply:
It is still broken

Reply:
Sorry for the mess, but now it should be ok to review again